src/main/kotlin/org/koitharu/kotatsu/parsers/site/TuMangaOnlineParser.kt

```diff
@@ -0,0 +1,382 @@
+				requestQueue.addLast(timestamp)
+			}
+		} finally {
+			fairLock.release()
+		}
+
+		val response = chain.proceed(request)
+		if (response.networkResponse == null) {
+			synchronized(requestQueue) {
+				if (requestQueue.isEmpty() || timestamp < requestQueue.first()) return@synchronized
+				val iterator = requestQueue.iterator()
+				while (iterator.hasNext()) {
+					if (iterator.next() == timestamp) {
+						iterator.remove()
+						break
+					}
+				}
+				(requestQueue as Object).notifyAll()
+			}
+		}
+
+		return response
+	}
+
+	override suspend fun getListPage(
+		page: Int,
+		query: String?,
+		tags: Set<MangaTag>?,
+		sortOrder: SortOrder
+	): List<Manga> {
+		val url = buildString {
+			append("/library")
+			if(query.isNullOrEmpty()){
+				append("?order_item=")
+				if (sortOrder == SortOrder.POPULARITY) {
+					append("likes_count")
+				}
+				if (sortOrder == SortOrder.NEWEST) {
+					append("creation")
+				}
+				append("&order_dir=desc")
+				append("&filter_by=title")
+				if (tags != null) {
+					for(tag in tags){
+						append("&genders[]=${tag.key}")
+					}
+				}
+			} else {
+				append("?title=$query")
+			}
+			append("&_pg=1")
+			append("&page=$page")
+		}.toAbsoluteUrl(domain)
+
+		val doc = webClient.httpGet(url, headers).parseHtml()
+		val items = doc.body().select("div.element")
+		return items.mapNotNull { item ->
+			val href = item.selectFirst("a")?.attrAsRelativeUrl("href")?.substringAfter(" ") ?: return@mapNotNull null
+			Manga(
+				id = generateUid(href),
+				title = item.selectFirst("h4.text-truncate")?.text() ?: return@mapNotNull null,
+				coverUrl = item.select("style").toString().substringAfter("('").substringBeforeLast("')"),
+				altTitle = null,
+				author = null,
+				rating = item.selectFirst("span.score")
+					?.text()
+					?.toFloatOrNull()
+					?.div(10F) ?: RATING_UNKNOWN,
+				url = href,
+				isNsfw = item.select("i").hasClass("fas fa-heartbeat fa-2x"),
+				tags = emptySet(),
+				state = null,
+				publicUrl = href.toAbsoluteUrl(doc.host ?: domain),
+				source = source,
+			)
+		}
+
+	}
+
+	override suspend fun getDetails(manga: Manga): Manga {
+		val doc = webClient.httpGet(manga.url.toAbsoluteUrl(domain)).parseHtml()
+		val contents = doc.body().selectFirstOrThrow("section.element-header-content")
+		return manga.copy(
+			description = contents.selectFirst("p.element-description")?.html(),
+			largeCoverUrl = contents.selectFirst(".book-thumbnail")?.attrAsAbsoluteUrlOrNull("src"),
+			state = parseStatus(contents.select("span.book-status").text().orEmpty()),
+			author = contents.selectFirst("h5.card-title")?.attr("title")?.substringAfter(", "),
+			chapters = if(doc.select("div.chapters").isEmpty()){
+				doc.select(oneShotChapterListSelector()).reversed().map { oneShotChapterFromElement(it) }
+			} else {
+				val chapters = mutableListOf<MangaChapter>()
+				doc.select(regularChapterListSelector()).reversed().forEachIndexed{ i, chapelement ->
```

Comment:
```suggestion
				doc.select(regularChapterListSelector()).mapChapters(reversed = true) { i, chapelement ->
```

Reply:
Here I'm not obtaining the MangaChapters directly; instead, I am adding them to a list (because each chapter can have multiple translations).